src/core/CONVENTIONS.md

```diff
@@ -0,0 +1,126 @@
+# Kibana Conventions
+
+- [Plugin Structure](#plugin-structure)
+
+## Plugin Structure
+
+All plugins built at Elastic should follow the same structure.
+
+```
+my_plugin/
+├── kibana.json
+├── public
+│   ├── applications
+│   │   └── my_app
+│   │       ├── components
+│   │       │   └── index.ts
+│   │       └── index.ts
+│   ├── index.ts
+│   ├── plugin.ts
+│   └── services
+│       └── index.ts
+└── server
+    ├── index.ts
+    ├── plugin.ts
+    ├── routes
+    │   └── index.ts
+    └── services
+        └── index.ts
+```
+
+- Both `server` and `public` should have an `index.ts` and a `plugin.ts` file:
+  - `index.ts` should only contain:
+    - The `plugin` export
+    - The `config` export (server only)
+    - Type exports for your Setup and Start interfaces, and any relevant sub-types
+    - Static, pure exports to be consumed by other plugins
+  - `plugin.ts` should only export a single class that implements the `Plugin` interface (more details below)
+- UI applications should live as modules inside the `applications` directory.
+  - Applications should export a single `renderApp` function.
+  - Applications should be loaded using dynamic async imports (more details below)
+  - Each application should contain a `components` subdirectory for storing individual components.
+- Services provided to other plugins as APIs should live inside the `services` subdirectory.
+- HTTP routes should be contained inside the `routes` directory.
+  - More should be fleshed out here...
+
+### The PluginInitializer
+
+```ts
+// my_plugin/public/index.ts
+
+import { PluginInitializer } from '../../src/core/public';
+import { MyPlugin, MyPluginSetup, MyPluginStart } from './plugin';
+
+export const plugin: PluginInitializer<MyPluginSetup, MyPluginStart> = () => new MyPlugin();
+export {
+  MyPluginSetup,
+  MyPluginStart
+}
+```
+
+### The Plugin class
+
+```ts
+// my_plugin/public/plugin.ts
+
+import { Plugin } from '../../src/core/public';
+
+export class MyPlugin implements Plugin {
+  public setup() { ... }
+  public start() { ... }
+  public stop() { ... }
```

Comment:
Are there "fairly stable" types defined for the arguments passed to these lifecycle methods (are they called lifecycle methods? idk)? It'd be fantastic to have the arguments and their types defined and linked to from here so we can get a sense of what we'll be doing in these methods. Thanks for this work!

Reply:
Yes there are types for this. You'll want to use the `CoreSetup` and `CoreStart` types from either src/core/public or src/core/server. By making your class implement the `Plugin` interface, you'll get an error if you accept any other type. Unfortunately, type interference doesn't assign the types automatically, so you'll have to do it yourself:


```ts
import { CoreSetup, CoreStart, Plugin } from '../../src/core/public';
export class MyPlugin implements Plugin {
  public setup(core: CoreSetup) { ... }
  public start(core: CoreStart) { ... }
  public stop() { ... } // does not accept any params at this time
}
```

These types are fairly stable, though we are still working on removing some from public's CoreSetup that should only be present in CoreStart. Other than that, the only major changes should be new services that are added.

Reply:
This is perfect! I'd just use that example in the conventions doc so folks can track down those Core* types if they are curious about what gets passed to those methods. :+1:

Reply:
Updated with additional detail!